### PR TITLE
Reference Implementation of Logical Reduce Operations

### PR DESCRIPTION
**Pull request by ggalieroc**
### Details: Reference Implementation of ReduceLogicalAnd and ReduceLogicalOr refactored.
 - Remove CoordinateTransform calls to index(). Benchmark results in 45979 (ReduceLogicalAnd) and 45975 (ReduceLogicalOr).
 - Enable negative axis values when axes are provided by host tensor.

### Tickets:
 - *ReduceLogicalAnd*: 37465, 45979
 - *ReduceLogicalOr*: 37466, 45975

### CODE REVIEW

**Comment by tomdol** [comment]
I was going to merge this https://github.com/openvinotoolkit/openvino/pull/5614 today after CF but it's been postponed until tomorrow. Can we merge mine first and then you could reuse the new axis extraction function I added? Or to avoid conflicts, you can sync your branch with master and even change the ReduceSum to the function you modified?

**Comment by ggalieroc** [comment]
Sure. I added you as a reviewer since there was already some existing helper which basically did the same but without support for negative axes. Other Reduce operation didn't handle this case (axis not a constant op) but Logical Reduce in particular did. You can merge yours first, then when resolving conflicts I can reuse your code, and remove the other remaining helper

File: ngraph/core/reference/include/ngraph/runtime/reference/logical_reduction.hpp [hunk 1]
```diff
@@ -17,53 +18,75 @@ namespace ngraph
...
+                        input_coord.begin(), input_coord.end(), in_strides.begin(), 0);
+                    size_t out_idx = std::inner_product(
+                        output_coord.begin(), output_coord.end(), out_strides.begin(), 0);
+
+                    out[out_idx] = out[out_idx] && arg[in_idx];
                 }
             }
 
+            // DEPRECATED, can't be removed currently due to arm-plugin dependency
+            static inline void reduce_logical_and(const char* arg,
```
Comment (ilyachur) [inline ngraph/core/reference/include/ngraph/runtime/reference/logical_reduction.hpp:47]: Can we use nGraph deprecated macro?

Comment (ggalieroc) [reply ngraph/core/reference/include/ngraph/runtime/reference/logical_reduction.hpp:47]: Done

File: ngraph/core/reference/include/ngraph/runtime/reference/logical_reduction.hpp [hunk 2]
```diff
@@ -17,53 +18,75 @@ namespace ngraph
...
                                                   char* out,
-                                                  const Shape& input_shape,
-                                                  const AxisSet& reduction_axes,
-                                                  bool keep_dims)
+                                                  const Shape& in_shape,
+                                                  const AxisSet& reduction_axes)
             {
-                CoordinateTransform output_transform(
-                    reduce(input_shape, reduction_axes, keep_dims));
+                constexpr bool dont_keep_dims_in_output = false;
```
Comment (tomdol) [inline ngraph/core/reference/include/ngraph/runtime/reference/logical_reduction.hpp:24]: The name of this variable is a little confusing to me. The consuming function `reduce()` takes it's value as `keep_dims` parameter. Should this variable be named the same way?

File: ngraph/core/src/op/reduce_logical_and.cpp [hunk 3]
```diff
@@ -62,16 +55,25 @@ namespace
...
                                         const HostTensorVector& inputs) const
 {
     NGRAPH_OP_SCOPE(v1_ReduceLogicalAnd_evaluate);
+    NGRAPH_CHECK(validate_host_tensor_vector(inputs, 2));
+    NGRAPH_CHECK(validate_host_tensor_vector(outputs, 1));
     const auto& data = inputs[0];
     const auto& axes = inputs[1];
     const auto& out = outputs[0];
-    return evaluate_reduce_logical_and(data, axes, out, get_keep_dims());
+    if (data->get_element_type() != element::boolean ||
```
Comment (tomdol) [inline ngraph/core/src/op/reduce_logical_and.cpp:69]: I think you can just use `is_integral()` here since it also returns true for booleans.

File: ngraph/core/src/op/reduce_logical_or.cpp [hunk 4]
```diff
@@ -62,16 +55,25 @@ namespace
...
                                        const HostTensorVector& inputs) const
 {
     NGRAPH_OP_SCOPE(v1_ReduceLogicalOr_evaluate);
+    NGRAPH_CHECK(validate_host_tensor_vector(inputs, 2));
+    NGRAPH_CHECK(validate_host_tensor_vector(outputs, 1));
     const auto& data = inputs[0];
     const auto& axes = inputs[1];
     const auto& out = outputs[0];
-    return evaluate_reduce_logical_or(data, axes, out, get_keep_dims());
+    if (data->get_element_type() != element::boolean ||
```
Comment (tomdol) [inline ngraph/core/src/op/reduce_logical_or.cpp:69]: Here as well - you can use `is_integral()` instead.

**Review by tomdol — CHANGES_REQUESTED** [review]

File: ngraph/core/reference/include/ngraph/runtime/reference/logical_reduction.hpp (hunk 2 above)
Comment (ggalieroc) [reply ngraph/core/reference/include/ngraph/runtime/reference/logical_reduction.hpp:24]: The prev implementation actually used `keep_dims` as this value could be `true` or `false`. But for index calculations we do not really need `keep_dims` attribute. Since resulting dimensions are 1, it does not affect the index. For this reason I created a constant expression as false to indicate that in ref impl we do not keep reduced dimensions in output. The name reflects the purpose of the variable. This change was suggested and merged in other ref implementations as ReduceSum and ReduceProd. Here original [thread](https://github.com/openvinotoolkit/openvino/pull/5728#discussion_r639640796).
I would rather keep it as it is now.

File: ngraph/core/src/op/reduce_logical_and.cpp (hunk 3 above)
Comment (ggalieroc) [reply ngraph/core/src/op/reduce_logical_and.cpp:69]: But checking `data ` input for `is_integral()` would allow integer element types for the input, which for logical reduce operations is forbidden based on current [spec ](https://github.com/openvinotoolkit/openvino/pull/5612/files#diff-2e4d9a4ca174d6a600cd347c59c0e695c63e257ef1e08778ddf08ba0b0fb810aR45)

File: ngraph/core/src/op/reduce_logical_or.cpp (hunk 4 above)
Comment (ggalieroc) [reply ngraph/core/src/op/reduce_logical_or.cpp:69]: Please check the previous comment in reduce logical and

File: ngraph/core/src/op/reduce_logical_and.cpp (hunk 3 above)
Comment (tomdol) [reply ngraph/core/src/op/reduce_logical_and.cpp:69]: Ok, in that case you should only check if `data->get_element_type() != element::boolean` in this `if`.

Comment (tomdol) [reply ngraph/core/src/op/reduce_logical_and.cpp:69]: My bad, the second check is for axes. You can disregard it, sorry for the noise ;)

**Review by tomdol — APPROVED** [review]

File: ngraph/core/reference/include/ngraph/runtime/reference/logical_reduction.hpp [hunk 5]
```diff
@@ -17,53 +18,75 @@ namespace ngraph
...
-                        arg[input_transform.index(input_coord)];
+                    const Coordinate output_coord =
+                        reduce(input_coord, reduction_axes, dont_keep_dims_in_output);
+
+                    const size_t in_idx = std::inner_product(
+                        input_coord.begin(), input_coord.end(), in_strides.begin(), 0);
+                    const size_t out_idx = std::inner_product(
+                        output_coord.begin(), output_coord.end(), out_strides.begin(), 0);
+
+                    out[out_idx] = out[out_idx] && arg[in_idx];
```
Comment (mitruska) [inline ngraph/core/reference/include/ngraph/runtime/reference/logical_reduction.hpp:42]: Would it be an overkill to reuse `reference::and`  implementation here? 
(same question for `or`)

**Review by mitruska — APPROVED** [review]

**Comment by ggalieroc** [comment]
@tomdol maybe you'll have time to review & merge to offload @ilyachur ? Unreviewed PR's started to pile up because of the release. CC: @andrei-kochin

File: ngraph/test/eval.cpp [hunk 6]
```diff
@@ -1810,28 +1810,6 @@ TEST(eval, topk_v1_param_dyn_k0)
     ASSERT_EQ(result1_val, expec1);
 }
 
-TEST(eval, reduce_logical_and__neg_axis)
```
Comment (ilyachur) [inline ngraph/test/eval.cpp:1813]: Why did you remove this test?

Comment (ggalieroc) [reply ngraph/test/eval.cpp:1813]: The purpose of the test was to throw exception when negative axis was introduced as second input. Since negative axes are supported I removed it

**Review by ilyachur — APPROVED** [review]
